### PR TITLE
Npm prep and modernization

### PR DESCRIPTION
**Pull request by Dan503**
I've done some major changes. I hope you don't mind.

1. Set up a gulp based build process to make development easier and allow for the use of modern JS.
2. Updated ES Lint config to allow for modern js
3. Converted driftory to modern class based syntax
4. Made the project ready for publishing to npm (haven't been able to fully test publish functionality though)
5. Made it so that you don't have to manually add the `openseadragon` script tag to the page. The js does it automatically.
6. Made it so that when you save a file, the demo site immediately refreshes automatically
7. Made the site get generated into a `docs` folder specifically for an easy GitHub pages set up
8. I renamed the project "driftory" because it's a much better name than "driftory-viewer".

To set up GitHub pages, do the following:

image

image

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -2,18 +2,17 @@
...
+    "ecmaVersion": 2020,
+    "sourceType": "module"
   },
   "globals": {
-    "OpenSeadragon": false,
-    "DriftoryViewer": true
+    "OpenSeadragon": true
   },
   "rules": {
-    "no-unused-vars": 0,
```
Comment (iangilman) [inline .eslintrc:16]: This one often bugs me, so I may reinstate it, but we can leave it out for now.

File: LICENCE [hunk 2]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (iangilman) [inline LICENCE]: Looks like this file is named incorrectly (should be LICENSE, right?).

**Review by iangilman — CHANGES_REQUESTED** [review]
Wow, this is fantastic! Thank you for doing all of this :-)

I'm of two minds on the naming (Driftory versus Driftory Viewer). I can see how it's nice to keep it short, but from a branding perspective, Driftory is the service, which includes all of the creation tools and the directory of comics and all that, whereas this library is just the viewer or player. 

By way of analogy, Driftory is like YouTube… A full-featured service with many aspects. If you take a YouTube video and place it on your own page, you wouldn't call it "a YouTube"; you'd call it a "YouTube embed", "YouTube player", "YouTube video" or something of that sort.

So that's why I went with "Driftory Viewer" for this library. It feels more accurate. What is your concern with the name?

Beyond that, this looks great (and thank you for the information on setting up GitHub pages for it).

File: LICENCE (hunk 2 above)
Comment (iangilman) [reply LICENCE]: Interesting… I just looked it up and discovered that "licence" is the correct spelling in Australia, so it's not a typo exactly! :-)

At any rate, I think GitHub wants it to be LICENSE:

https://docs.github.com/en/rest/reference/licenses

File: .eslintrc (hunk 1 above)
Comment (Dan503) [reply .eslintrc:16]: unused vars tends to lead to importing libraries that you aren't actually using (increasing bundle size) or having lots of code that isn't actually being used so I'd like to keep no-unused-vars turned on.

File: docs/demo.css [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
-  box-sizing: border-box;
+  /* 1rem = 10px making font size calculations easier */
+  font-size: 62.5%;
 }
-*, *:before, *:after {
-  box-sizing: inherit;
+*,
+*:before,
+*:after {
+  box-sizing: border-box;
```
Comment (Dan503) [inline docs/demo.css:9]: I didn't really look at the CSS but I'm noticing this now.

I've tried both the `* {box-sizing: inherit;}` method and the `* { box-sizing: border-box;}` method in real projects.

I've found that I much rather the `* { box-sizing: border-box;}` method. It allows you to make a parent container `box-sizing: content box` without having to re-apply `box-sizing: border-box` to child elements.

The one use case where the inherit model is better is when you have an old 3rd party plugin or something that you don't control the styling for. In those cases you can do this:

```css
.inherit-content-box-sizing {
  box-sizing: content-box;
}
.inherit-content-box-sizing *,
.inherit-content-box-sizing *::before,
.inherit-content-box-sizing *::after {
  box-sizing: inherit;
}
```

File: LICENCE (hunk 2 above)
Comment (Dan503) [reply LICENCE]: Changed the file name to the American spelling.

**Comment by Dan503** [comment]
I'm glad you are happy with the changes :)

I was a bit nervous since it is a bit of a departure from what you are probably used to.

As for the name, it's a lot better from a marketing perspective to just call it "driftory". It's short and catchy.

I get what you mean by it not technically being `"driftory"` though. There is technically [a "youtube" npm module](https://www.npmjs.com/package/youtube) but it is far from being official.

As a developer ingesting the API, I don't really make the distinction between "this is just the viewer, the website is driftory itself". When using it, I see it more like this:

> Ok, I'm going to create a driftory package using the website. Now I'll use the official "driftory" package to run it. 
>
> `npm i driftory` ... error ... why didn't that work? 😢 .... oh it's called `driftory-viewer` not `driftory`

I would also prefer to use `new Driftory({ ... })` when initializing it in my code than `new DriftoryViewer({ ... })` though since we are using a default export the developer can call it whatever they want when importing it.

If you want to stick with calling it "driftory-viewer" to be more technically correct then it's ultimately your call to make but I would personally prefer it to be called just "driftory".

**Comment by Dan503** [comment]
@iangilman I've come up with another idea that I think will satisfy both of us.

1. You create a personal account on npm
2. Create an organization under your account called "driftory"

image

3. We name this package `@driftory/viewer`
4. Publish using `npm publish --access public`
5. Add me as a member of the organization so that I can also publish package versions.

There are a fair few benefits to doing this:

1. This tells the developers using it that "driftory" is the organization name, not the package name.
2. If there are any other driftory related things you want to build in future that you want to publish to npm  (eg. a Driftory generator maybe?), you don't have to worry about the name getting taken. You publish everything under the `@driftory` name space which is locked to your account.
3. It gives me the ability to also publish the package without me having to juggle different npm accounts.
4. You get to keep it technically correct only calling it the viewer rather than just "driftory".

File: .eslintrc (hunk 1 above)
Comment (iangilman) [reply .eslintrc:16]: Seems like that would only apply to unused imports, right? Is there a specific flag for that?

File: docs/demo.css (hunk 3 above)
Comment (iangilman) [reply docs/demo.css:9]: Cool… Works for me :-)

**Review by iangilman — APPROVED** [review]
Using an npm org sounds like a great idea! I'll work on getting that set up. In the meantime, hopefully you can make use of this repository so you're not blocked.

Of course I'll also work on bringing the viewer up to spec.

> I was a bit nervous since it is a bit of a departure from what you are probably used to.

I'm familiar with Babel and ES6, but not Gulp and Rollup (I'm more familiar with Grunt and Webpack). Always great to learn new stuff, though! :-)

I'm going to go ahead and merge this; feel free to make follow-up pull requests as needed…

**Comment by iangilman** [comment]
It's deployed now to GitHub pages:

http://pixfabrik.com/driftory-viewer/

… But it's broken. I'll take a look.

File: .eslintrc (hunk 1 above)
Comment (Dan503) [reply .eslintrc:16]: Not necessarily. There are times where you might be refactoring code and then suddenly a whole chain of functions that were all dependent on one another aren't used anymore so their presence in the code base is just causing code bloat and making the code base harder to maintain.

The presence of unused variables tends to signify that bad coding practices are being used. In other words, it's a bit of a "code smell'.

**Comment by Dan503** [comment]
> I'm familiar with Babel and ES6, but not Gulp and Rollup (I'm more familiar with Grunt and Webpack). Always great to learn new stuff, though! :-)

You can think of Gulp as basically just Grunt but way way waaaaay better.
It's faster than Grunt since it can run tasks in parallel. Grunt in my experience can only run tasks sequentially. Gulp also has a much nicer configuration syntax that is easier to understand what is going on and also easier to write and maintain. You can make some pretty awesome automation happen with Gulp 😃

**Comment by Dan503** [comment]
I changed the package name to `"@driftory/viewer"` so it should be ready for a test publish now once you have set up the organisation and fixed the issue with github pages.

**Comment by iangilman** [comment]
Cool, thanks! I'll let you know once I've pushed it (just as soon as I get a spare moment).

**Pull request merged**